### PR TITLE
Add TransientMemoryTracker.

### PR DESCRIPTION
**Pull request by toregge**
@vekterli : please review

### CODE REVIEW

File: vespalib/src/tests/util/transient_memory_tracker_test.cpp [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    EXPECT_EQ(0, totals._total_transient_memory);
+    EXPECT_EQ(_initial_totals._generation, totals._generation);
+}
+
+TEST_F(TransientMemoryTrackerTest, normal_transient_memory_lifetime) {
+    TransientMemoryTracker transient_memory_tracker;
+    set_transient_memory(transient_memory_tracker, 42);
+    auto totals = TransientMemoryTracker::get_total_transient_memory();
+    EXPECT_EQ(42, totals._total_transient_memory);
+    EXPECT_NE(_initial_totals._generation, totals._generation);
```
Comment (vekterli) [inline vespalib/src/tests/util/transient_memory_tracker_test.cpp]: Consider using `EXPECT_LT` if we want to check strictly increasing generations (also applies to other generation comparisons)

File: vespalib/src/vespa/vespalib/util/transient_memory_tracker.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+#include <cstdint>
+#include <mutex>
+
+namespace vespalib {
+
+/*
+ * Class used to track large transient memory allocations, e.g. when flushing
+ * attribute vectors.
+ */
```
Comment (vekterli) [inline vespalib/src/vespa/vespalib/util/transient_memory_tracker.h:18]: Consider adding a brief comment about the intended use case of the generation tracking

File: vespalib/src/vespa/vespalib/util/transient_memory_tracker.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        size_t   _total_transient_memory;
+        uint64_t _generation;
+    };
+    TransientMemoryTracker() noexcept;
+    TransientMemoryTracker(const TransientMemoryTracker&) = delete;
+    TransientMemoryTracker(TransientMemoryTracker&& rhs) noexcept;
+    ~TransientMemoryTracker();
+    TransientMemoryTracker& operator=(const TransientMemoryTracker&) = delete;
+    TransientMemoryTracker& operator=(TransientMemoryTracker&& rhs) noexcept;
+    static Lock get_lock() { return Lock(_mutex); }
```
Comment (vekterli) [inline vespalib/src/vespa/vespalib/util/transient_memory_tracker.h]: Consider calling this `acquire_lock` instead, as this could otherwise be confused with a method that returns a ref to the underlying _mutex_ rather than returning a `unique_lock` by-value for that mutex.

Also consider `[[nodiscard]]`

File: vespalib/src/vespa/vespalib/util/transient_memory_tracker.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    void set_transient_memory(size_t value);
+    void set_transient_memory(Lock lock, size_t value);
+    void swap(TransientMemoryTracker& rhs) noexcept;
+    static TotalTransientMemoryAndGeneration get_total_transient_memory();
```
Comment (vekterli) [inline vespalib/src/vespa/vespalib/util/transient_memory_tracker.h]: Consider `[[nodiscard]]` (and possibly also `noexcept` to indicate this can always safely be sampled)

Comment (vekterli) [inline vespalib/src/vespa/vespalib/util/transient_memory_tracker.h]: Consider adding a comment that this must _not_ be called while holding the lock

File: vespalib/src/vespa/vespalib/util/transient_memory_tracker.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    assert(lock.mutex() == &_mutex);
+    assert(lock.owns_lock());
+    if (value == _transient_memory) {
+        return;
+    }
+    if (value > _transient_memory) {
+        _total_transient_memory += value - _transient_memory;
+    } else if (value < _transient_memory) {
+        _total_transient_memory -= _transient_memory - value;
+    }
```
Comment (vekterli) [inline vespalib/src/vespa/vespalib/util/transient_memory_tracker.cpp:52]: Consider adding a test where these branches are exercised with a non-zero from/to value to show that they are incremental as expected.

**Review by vekterli — DISMISSED** [review]
:+1:

**Review by vekterli — APPROVED** [review]
:+1:
